federatedrec/param/svdpp_param.py

```diff
@@ -0,0 +1,271 @@
+                        self.optimizer.kwargs['learning_rate'] < 0):
+                raise ValueError(
+                    "svdpp's optimizer.learning_rate {} not supported, should be 'int' or 'float', "
+                    "and greater than 0".format(
+                        self.optimizer.kwargs['learning_rate']))
+
+        return True
+
+    def _parse_early_stop(self, param):
+        """
+           Examples:
+
+               1. "early_stop": "diff"
+               2. "early_stop": {
+                       "early_stop": "diff",
+                       "eps": 0.0001
+                   }
+        """
+        default_eps = 0.0001
+        if isinstance(param, str):
+            return SimpleNamespace(converge_func=param, eps=default_eps)
+        elif isinstance(param, dict):
+            early_stop = param.get("early_stop", None)
+            eps = param.get("eps", default_eps)
+            if not early_stop:
+                raise ValueError(f"early_stop config: {param} invalid")
+            return SimpleNamespace(converge_func=early_stop, eps=eps)
+        else:
+            raise ValueError(f"invalid type for early_stop: {type(param)}")
+
+    def _parse_optimizer(self, param):
+        """
+        Examples:
+
+            1. "optimize": "SGD"
+            2. "optimize": {
+                    "optimizer": "SGD",
+                    "learning_rate": 0.05
+                }
+        """
+        kwargs = {}
+        if isinstance(param, str):
+            return SimpleNamespace(optimizer=param, kwargs=kwargs)
+        elif isinstance(param, dict):
+            optimizer = param.get("optimizer", kwargs)
+            if not optimizer:
+                raise ValueError(f"optimizer config: {param} invalid")
+            kwargs = {k: v for k, v in param.items() if k != "optimizer"}
+            return SimpleNamespace(optimizer=optimizer, kwargs=kwargs)
+        else:
+            raise ValueError(f"invalid type for optimize: {type(param)}")
+
+    def _parse_metrics(self, param):
+        """
+        Examples:
+
+            1. "metrics": "Accuracy"
+            2. "metrics": ["Accuracy"]
+        """
+        if not param:
+            return []
+        elif isinstance(param, str):
+            return [param]
+        elif isinstance(param, list):
+            return param
+        else:
+            raise ValueError(f"invalid metrics type: {type(param)}")
+
+
+class HeteroSVDppParam(SVDppParam):
+    """
+    Parameters
+    ----------
+    re_encrypt_batches : int, default: 2
+        Required when using encrypted version HomoLR. Since multiple batch updating coefficient may cause
+        overflow error. The model need to be re-encrypt for every several batches. Please be careful when setting
+        this parameter. Too large batches may cause training failure.
+
+    aggregate_iters : int, default: 1
+        Indicate how many iterations are aggregated once.
+
+    """
+
+    def __init__(self, optimizer='sgd',
```

Comment:
When using tf's optimizer, it will be case sensitive. The default value optimizer='sgd' may raise some error?